panic on windows (version string "v" doesn't match expected regular expression: "^v(\d+\.\d+\.\d+)"

### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Version

1.24.2

### Platform

```shell
windows server 2019
```


### What happened?

When I call k0s.exe, no matter the arguments I get this error:

k0s worker --cri-socket=docker:tcp://127.0.0.1:2375 --cidr-range=10.0.0.10/24 --cluster-dns=1.2.3.4 --api-server=http://batata asgsdfgsde
panic: version string "v" doesn't match expected regular expression: "^v(\d+\.\d+\.\d+)"

goroutine 1 [running]:
k8s.io/component-base/metrics.parseVersion({{0x3c03890, 0x0}, {0x3c03891, 0x0}, {0x3c038ae, 0x1}, {0x3c03fd0, 0xd}, {0x0, 0x0}, ...})
        /run/k0s-build/go/mod/k8s.io/component-base@v0.24.2/metrics/version_parser.go:47 +0x274
k8s.io/component-base/metrics.newKubeRegistry({{0x3c03890, 0x0}, {0x3c03891, 0x0}, {0x3c038ae, 0x1}, {0x3c03fd0, 0xd}, {0x0, 0x0}, ...})
        /run/k0s-build/go/mod/k8s.io/component-base@v0.24.2/metrics/registry.go:320 +0x118
k8s.io/component-base/metrics.NewKubeRegistry()
        /run/k0s-build/go/mod/k8s.io/component-base@v0.24.2/metrics/registry.go:335 +0x78
k8s.io/component-base/metrics/legacyregistry.init()
        /run/k0s-build/go/mod/k8s.io/component-base@v0.24.2/metrics/legacyregistry/registry.go:29 +0x1d

### Steps to reproduce

1. download the amd64.exe and run it on windows
2.
3.


### Expected behavior

It should work

### Actual behavior

It panics and abort

### Screenshots and logs

_No response_

### Additional context

_No response_

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -58,15 +58,15 @@ jobs:
       - name: Prepare build environment
         run: .github/workflows/prepare-build-env.sh
 
-      - name: Set up Go
-        uses: actions/setup-go@v3
-        with:
-          go-version: ${{ env.GO_VERSION }}
-
       - name: Build
         run: make EMBEDDED_BINS_BUILDMODE=docker
         env:
           VERSION: ${{ needs.release.outputs.tag_name }}
+
+      - name: Set up Go for smoke tests
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
 
       - name: Run basic smoke test
         run: make check-basic
@@ -108,12 +108,12 @@ jobs:
           asset_path: ./image-bundle/bundle.tar
           asset_name: k0s-airgap-bundle-${{ needs.release.outputs.tag_name }}-amd64
           asset_content_type: application/octet-stream
-      
+
       - name: Clean Docker after build
         if: always()
         run: |
           docker system prune --all --volumes --force
-      
+
       # https://github.com/actions/checkout/issues/273#issuecomment-642908752
       # Golang mod cache tends to set directories to read-only, which breaks any
       # attempts to simply remove those directories. The `make clean-gocache`
@@ -193,15 +193,15 @@ jobs:
         run: .github/workflows/prepare-build-env.sh
         working-directory: ./
 
-      - name: Set up Go
-        uses: actions/setup-go@v3
-        with:
-          go-version: ${{ env.GO_VERSION }}
-
       - name: Build
         run: make EMBEDDED_BINS_BUILDMODE=docker
         env:
           VERSION: ${{ needs.release.outputs.tag_name }}
+
+      - name: Set up Go for smoke tests
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
 
       - name: Run basic smoke test
         run: make check-basic
@@ -271,18 +271,35 @@ jobs:
       - name: "Pre: Fixup directories"
         run: find . -type d -not -perm /u+w -exec chmod u+w '{}' \;
 
-      - name: Install GoLang for ARMHF
-        run: "echo $HOME/.local/go/bin >> $GITHUB_PATH; rm -rf $HOME/.local/go && mkdir -p $HOME/.local/go && curl --silent -L https://golang.org/dl/$(curl --silent -L 'https://golang.org/VERSION?m=text').linux-armv6l.tar.gz | tar -C $HOME/.local -xz"
-      - name: Go Version
-        run: go version
-
       - name: Check out code into the Go module directory
         uses: actions/checkout@v3
+
+      - name: Prepare build environment
+        run: .github/workflows/prepare-build-env.sh
+        working-directory: ./
 
       - name: Build
         run: make EMBEDDED_BINS_BUILDMODE=docker
         env:
           VERSION: ${{ needs.release.outputs.tag_name }}
+
+      # Need to install Go manually: https://github.com/actions/setup-go/issues/106
+      - name: Set up Go for smoke tests (armv6l)
+        run: |
+          echo "Setup go stable version $GO_VERSION"
+          rm -rf -- "$HOME/.local/go"
+          mkdir -p -- "$HOME/.local/go"
+          curl --silent -L "https://golang.org/dl/go${GO_VERSION}.linux-armv6l.tar.gz" | tar -C "$HOME/.local" -xz
+
+          echo "$HOME/.local/go/bin" >>"$GITHUB_PATH"
+          export PATH="$PATH:$HOME/.local/go/bin"
+          echo Added go to the path
+
+          echo "Successfully setup go version $GO_VERSION"
+          go version
+          echo ::group::go env
+          go env
+          echo ::endgroup::
 
       - name: Run basic smoke test
         run: make check-basic
@@ -389,7 +406,6 @@ jobs:
             docker.io/k0sproject/k0s:${{ needs.release.outputs.image_tag }}
           push: true
 
-
   conformance-test:
     needs:
       - release
@@ -470,4 +486,3 @@ jobs:
 
           terraform destroy -auto-approve
         if: ${{ always() }}
-

--- a/Makefile
+++ b/Makefile
@@ -25,10 +25,6 @@ ifeq ($(DEBUG), false)
 LD_FLAGS ?= -w -s
 endif
 
-KUBECTL_VERSION = $(shell go mod graph |  grep "github.com/k0sproject/k0s" |  grep kubectl  | cut -d "@" -f 2 | sed "s/v0\./1./")
-KUBECTL_MAJOR= $(shell echo ${KUBECTL_VERSION} | cut -d "." -f 1)
-KUBECTL_MINOR= $(shell echo ${KUBECTL_VERSION} | cut -d "." -f 2)
-
 # https://reproducible-builds.org/docs/source-date-epoch/#makefile
 # https://reproducible-builds.org/docs/source-date-epoch/#git
 # https://stackoverflow.com/a/15103333
@@ -48,9 +44,9 @@ LD_FLAGS += -X github.com/k0sproject/k0s/pkg/build.EtcdVersion=$(etcd_version)
 LD_FLAGS += -X github.com/k0sproject/k0s/pkg/build.KonnectivityVersion=$(konnectivity_version)
 LD_FLAGS += -X "github.com/k0sproject/k0s/pkg/build.EulaNotice=$(EULA_NOTICE)"
 LD_FLAGS += -X github.com/k0sproject/k0s/pkg/telemetry.segmentToken=$(SEGMENT_TOKEN)
-LD_FLAGS += -X k8s.io/component-base/version.gitVersion=v$(KUBECTL_VERSION)
-LD_FLAGS += -X k8s.io/component-base/version.gitMajor=$(KUBECTL_MAJOR)
-LD_FLAGS += -X k8s.io/component-base/version.gitMinor=$(KUBECTL_MINOR)
+LD_FLAGS += -X k8s.io/component-base/version.gitVersion=v$(kubernetes_version)
+LD_FLAGS += -X k8s.io/component-base/version.gitMajor=$(shell echo '$(kubernetes_version)' | cut -d. -f1)
+LD_FLAGS += -X k8s.io/component-base/version.gitMinor=$(shell echo '$(kubernetes_version)' | cut -d. -f2)
 LD_FLAGS += -X k8s.io/component-base/version.buildDate=$(BUILD_DATE)
 LD_FLAGS += -X k8s.io/component-base/version.gitCommit=not_available
 LD_FLAGS += -X github.com/containerd/containerd/version.Version=$(containerd_version)

--- a/go.mod
+++ b/go.mod
@@ -50,6 +50,7 @@ require (
 	golang.org/x/crypto v0.0.0-20220525230936-793ad666bf5e
 	golang.org/x/sync v0.0.0-20220601150217-0de741cfad7f
 	golang.org/x/sys v0.0.0-20220422013727-9388b58f7150
+	golang.org/x/tools v0.1.10-0.20220218145154-897bd77cd717
 	google.golang.org/grpc v1.48.0
 	gopkg.in/yaml.v2 v2.4.0
 	helm.sh/helm/v3 v3.9.1
@@ -262,7 +263,6 @@ require (
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/text v0.3.7 // indirect
 	golang.org/x/time v0.0.0-20220411224347-583f2d630306 // indirect
-	golang.org/x/tools v0.1.10-0.20220218145154-897bd77cd717 // indirect
 	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
 	gomodules.xyz/jsonpatch/v2 v2.2.0 // indirect
 	google.golang.org/appengine v1.6.7 // indirect

--- a/pkg/constant/constant_shared_test.go
+++ b/pkg/constant/constant_shared_test.go
@@ -23,22 +23,107 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
+	"golang.org/x/tools/go/packages"
 )
 
-func TestKonnectivityVersion(t *testing.T) {
-	assert.Equal(t, getVersion(t, "konnectivity"), strings.TrimPrefix(KonnectivityImageVersion, "v"))
+func TestConstants(t *testing.T) {
+	for _, test := range []struct{ name, constant, varName string }{
+		{"KonnectivityImageVersion", "v" + KonnectivityImageVersion, "konnectivity"},
+		{"KubeProxyImageVersion", KubeProxyImageVersion, "kubernetes"},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			assert.Equal(t, "v"+getVersion(t, test.varName), test.constant)
+		})
+	}
+
+	t.Run("KubernetesMajorMinorVersion", func(t *testing.T) {
+		ver := strings.Split(getVersion(t, "kubernetes"), ".")
+		require.GreaterOrEqual(t, len(ver), 2, "failed to spilt Kubernetes version %q", ver)
+		kubeMajorMinor := ver[0] + "." + ver[1]
+		assert.Equal(t, kubeMajorMinor, KubernetesMajorMinorVersion)
+	})
 }
 
-func TestKubeProxyVersion(t *testing.T) {
-	assert.Equal(t, getVersion(t, "kubernetes"), strings.TrimPrefix(KubeProxyImageVersion, "v"))
+func TestKubernetesModuleVersions(t *testing.T) {
+	kubernetesVersion := getVersion(t, "kubernetes")
+
+	checkPackageModules(t,
+		func(modulePath string) bool {
+			switch modulePath {
+			// Don't report any version mismatches on the following modules.
+			// They have a release cycle which is decoupled from k8s itself.
+			case "k8s.io/klog/v2", "k8s.io/kube-openapi", "k8s.io/utils":
+				return false
+
+			default:
+				return strings.HasPrefix(modulePath, "k8s.io/")
+			}
+		},
+		func(t *testing.T, pkgPath string, module *packages.Module) bool {
+			modVer := module.Version
+			if module.Path != "k8s.io/kubernetes" {
+				// All modules besides Kubernetes itself use v0 instead of v1.
+				modVer = strings.Replace(modVer, "v0.", "v1.", 1)
+			}
+
+			return !assert.Equal(t, "v"+kubernetesVersion, modVer,
+				"Module version for package %s doesn't match: %+#v",
+				pkgPath, module,
+			)
+		},
+	)
 }
 
-func TestKubernetesMajorMinorVersion(t *testing.T) {
-	ver := strings.Split(getVersion(t, "kubernetes"), ".")
-	require.GreaterOrEqual(t, len(ver), 2, "failed to spilt Kubernetes version %q", ver)
-	kubeMinorMajor := ver[0] + "." + ver[1]
+func TestEtcdModuleVersions(t *testing.T) {
+	etcdVersion := getVersion(t, "etcd")
+	etcdVersionParts := strings.Split(etcdVersion, ".")
+	require.GreaterOrEqual(t, len(etcdVersionParts), 1, "failed to spilt etcd version %q", etcdVersion)
 
-	assert.Equal(t, kubeMinorMajor, KubernetesMajorMinorVersion)
+	checkPackageModules(t,
+		func(modulePath string) bool {
+			return strings.HasPrefix(modulePath, "go.etcd.io/etcd/") &&
+				strings.HasSuffix(modulePath, "/v"+etcdVersionParts[0])
+		},
+		func(t *testing.T, pkgPath string, module *packages.Module) bool {
+			return !assert.Equal(t, "v"+etcdVersion, module.Version,
+				"Module version for package %s doesn't match: %+#v",
+				pkgPath, module,
+			)
+		},
+	)
+}
+
+func TestContainerdModuleVersions(t *testing.T) {
+	containerdVersion := getVersion(t, "containerd")
+
+	checkPackageModules(t,
+		func(modulePath string) bool {
+			return modulePath == "github.com/containerd/containerd"
+		},
+		func(t *testing.T, pkgPath string, module *packages.Module) bool {
+			return !assert.Equal(t, "v"+containerdVersion, module.Version,
+				"Module version for package %s doesn't match: %+#v",
+				pkgPath, module,
+			)
+		},
+	)
+}
+
+func TestRuncModuleVersions(t *testing.T) {
+	runcVersion := getVersion(t, "runc")
+
+	checkPackageModules(t,
+		func(modulePath string) bool {
+			return modulePath == "github.com/opencontainers/runc"
+		},
+		func(t *testing.T, pkgPath string, module *packages.Module) bool {
+			return !assert.Equal(t, "v"+runcVersion, module.Version,
+				"Module version for package %s doesn't match: %+#v",
+				pkgPath, module,
+			)
+		},
+	)
 }
 
 func getVersion(t *testing.T, component string) string {
@@ -50,4 +135,35 @@ func getVersion(t *testing.T, component string) string {
 	require.NotEmpty(t, out, "failed to get %s version", component)
 
 	return strings.TrimSuffix(string(out), "\n")
+}
+
+func checkPackageModules(t *testing.T, filter func(modulePath string) bool, check func(t *testing.T, pkgPath string, module *packages.Module) bool) {
+	pkgs, err := packages.Load(&packages.Config{
+		Mode: packages.NeedName | packages.NeedModule | packages.NeedImports | packages.NeedDeps,
+		Logf: t.Logf,
+	}, "github.com/k0sproject/k0s")
+	require.NoError(t, err)
+
+	failedModules := make(map[string]bool)
+	checkCalledAtLeastOnce := false
+
+	packages.Visit(pkgs, func(p *packages.Package) bool {
+		if p.Module != nil && filter(p.Module.Path) {
+			actual := p.Module
+			for actual.Replace != nil {
+				actual = actual.Replace
+			}
+
+			if !failedModules[actual.Path] {
+				checkCalledAtLeastOnce = true
+				if !check(t, p.PkgPath, actual) {
+					failedModules[actual.Path] = true
+				}
+			}
+		}
+
+		return true
+	}, nil)
+
+	assert.True(t, checkCalledAtLeastOnce, "Not a single package passed the filter.")
 }

